OverlayRegions should take clipping into account

etiennesegonzac opened the pull request:
#### dae59b416824a7e30ab2d6378603905188aa9c64
<pre>
OverlayRegions should take clipping into account
<a href="https://bugs.webkit.org/show_bug.cgi?id=295588">https://bugs.webkit.org/show_bug.cgi?id=295588</a>
&lt;<a href="https://rdar.apple.com/135734728">rdar://135734728</a>&gt;

Reviewed by Simon Fraser.

Overlay Regions are not part of the compositing tree (they&apos;re just
passed along as Rects). This doesn&apos;t work well in cases where we size the
EventRegion based on the layer&apos;s bounds.

Take clipping and masking into account while configuring a ScrollView
with overlay regions.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(configureScrollViewWithOverlayRegionsIDs):
Apply clipping and masking to the Region as we walk up the overlay&apos;s
ancestors.

* LayoutTests/overlay-region/clipping-expected.txt: Added.
* LayoutTests/overlay-region/clipping.html: Added.
New test covering clipping and masking scenarios.
* LayoutTests/overlay-region/snapping-expected.txt:
Updated expectations now that the scroller clips.

Canonical link: <a href="https://commits.webkit.org/297155@main">https://commits.webkit.org/297155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36ec19d8330643599076af63728b401576a35e99

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110605 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20697 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116631 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60872 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30943 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38853 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84112 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113553 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24713 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99596 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64552 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24076 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17736 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60426 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94099 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119421 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37645 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27960 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93077 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38019 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92900 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23693 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37913 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15656 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33622 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37540 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43011 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37203 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40542 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38911 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/36ec19d8330643599076af63728b401576a35e99. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47718

etiennesegonzac committed:
OverlayRegions should take clipping into account
https://bugs.webkit.org/show_bug.cgi?id=295588
<rdar://135734728>

Reviewed by Simon Fraser.

Overlay Regions are not part of the compositing tree (they're just
passed along as Rects). This doesn't work well in cases where we size the
EventRegion based on the layer's bounds.

Take clipping and masking into account while configuring a ScrollView
with overlay regions.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(configureScrollViewWithOverlayRegionsIDs):
Apply clipping and masking to the Region as we walk up the overlay's
ancestors.

* LayoutTests/overlay-region/clipping-expected.txt: Added.
* LayoutTests/overlay-region/clipping.html: Added.
New test covering clipping and masking scenarios.
* LayoutTests/overlay-region/snapping-expected.txt:
Updated expectations now that the scroller clips.

Canonical link: https://commits.webkit.org/297155@main
